domain/permission/state/state.go

```diff
@@ -356,3 +222,165 @@ func (s *State) ReadAllUserAccessForTarget(ctx context.Context, target corepermi
+	// id of spec.Target.ObjectType from permission_object_type as object_type_id
+	// Use access_type_id and object_type_id to validate row from permission_object_access
+	objectAccessIDExists := `
+SELECT permission_access_type.id AS &M.access_type_id
+FROM permission_object_access
+LEFT JOIN permission_object_type
+	ON permission_object_type.type = $M.object_type
+LEFT JOIN permission_access_type
+	ON permission_access_type.type = $M.access_type
+WHERE permission_object_access.access_type_id = permission_access_type.id
+	AND permission_object_access.object_type_id = permission_object_type.id
+`
+
+	// Validate the access type is allowed for the target type.
+	objectAccessIDStmt, err := sqlair.Prepare(objectAccessIDExists, sqlair.M{})
+	if err != nil {
+		return -1, errors.Trace(err)
+	}
+
+	var resultM = sqlair.M{}
+	err = tx.Query(ctx, objectAccessIDStmt, sqlair.M{"access_type": access, "object_type": objectType}).Get(&resultM)
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
+		return -1, errors.Annotatef(err, "mismatch in %q, %q", access, objectType)
+	} else if err != nil {
+		return -1, errors.Annotatef(err, "getting id for pair %q, %q", access, objectType)
+	}
+	return resultM["access_type_id"].(int64), nil
+}
+
+// validateTargetExists validates that the target of the permission
+// exists. An error is returned if not. Unless we have a controller
+// target, search for grant_on as a cloud.name and a model_list.uuid.
+// It must be one of those.
+func validateTargetExists(
+	ctx context.Context,
+	tx *sqlair.TX,
+	targetKey string,
+) error {
+	if targetKey == coredatabase.ControllerNS {
+		return nil
+	}
+
+	// TODO: (hml) 6-Mar-24
+	// Add application offers check here when added to DDL.
+	targetExists := `
+SELECT &M.found_it FROM (
+    SELECT 1 AS found_it FROM cloud WHERE cloud.name = $M.grant_on
+    UNION
+    SELECT 1 AS found_it FROM model_list WHERE model_list.uuid = $M.grant_on
+)
+`
+	// Validate the grant_on target exists.
+	targetExistsStmt, err := sqlair.Prepare(targetExists, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Check that 1 row exists if the grant_on value is not ControllerNS.
+	// The behavior for this check is changing in sqlair, trying to
+	// account for either behavior. Old behavior: return no error and
+	// a slice length of zero. New behavior: if the length is 0, return
+	// ErrNoRows.
+	var foundIt = []sqlair.M{}
+	err = tx.Query(ctx, targetExistsStmt, sqlair.M{"grant_on": targetKey}).GetAll(&foundIt)
+	if err != nil {
+		return errors.Annotatef(err, "verifying %q target exists", targetKey)
+	}
+
+	if len(foundIt) == 1 {
+		return nil
+	}
+
+	// Any answer other than 1 is an error. The targetKey should exist
+	// as a unique identifier across the controller namespace.
+	if len(foundIt) == 0 {
+		return fmt.Errorf("%q %w", targetKey, permissionerrors.TargetInvalid)
+	}
+	return fmt.Errorf("%q %w", targetKey, permissionerrors.TargetAlreadyExists)
```

Comment:
This error doesn't make sense here. The function is trying to determine if the target exists. The last failure is a should never happen, the UUID is not actually unique, or we've added 2 clouds with the same name. 

Nor am I sure it's a necessary type, as we have permission errors.AlreadyExists. Permissions is never adding a target.

Reply:
UniqueIdentifierIsNotUnique?

Reply:
Sure I'm happy with that. Wanna make the change and merge?